CONTRIBUTING.md

```diff
@@ -75,6 +90,133 @@ If the change is needed for a bigger feature, make that feature a separate pull
+As an example:
+
+```
+Fixed execresult function
+```
+
+This is **not good**, because it doesn't really say what was wrong.
+Here are some better examples:
+
+```
+Policy function execresult() no longer crashes when passed empty string
+```
+
+```
+Fixed memory leak in policy function execresult()
+```
+
+
+#### Commit message example
+
+Here is an example of our commit messages:
+
+```
+Fixed memory leak in: differences(), intersection(), unique()
+
+They all use `FnCallSetOp()` internally, which was lacking free
+for some error handling return paths.
+
+Changelog: Title
+Ticket: ENT-4586
+```
+
+A commit message consists of 3 parts; a title, a body, and metadata.
+The 3 parts are separated by empty lines.
+The body and metadata parts are optional.
+
+The first line of the commit message is called the title.
+It should not have punctuation (like `.`/`!`) on the end of the line.
+It should not exceed 80 characters.
+In a short sentence it explains the change in the commit.
+If the commit includes a change in code, as well as a test, it doesn't need to mention the test.
+Past tense is preferred, especially when the title is used in the changelog.
+
+The body can be several sentences.
+They should have proper punctuation, ending sentences in `.`/`!`.
+Wrap lines before 80 characters, so the width of `git log` is consistent.
+It should explain the change in detail, and can give examples, implementation details, etc.
+
+Using some markdown syntax is okay, but not required.
+Commit messages and changelog entries are considered plain text.
+Sometimes markdown can make them more clear, and the automatic pull request description on GitHub looks nicer.
+
+The metadata part is one or more lines of `Keyword: Value`.
+We mainly use 3 keywords:
+
+```
+Changelog: Title
+Ticket: CFE-1234
+Signed-off-by: Ole Herman Schumacher Elgesem <ole@northern.tech>
+```
+
+(`Signed-off-by` is added by `git commit -s`).
+
+One exception is the line automatically added by `git cherry-pick -x`, which looks like this:
+
+```
+(cherry picked from commit fca949fdb4b9691338e316b3675d8e5a4e9082b9)
+```
+
+If your change only affects one feature, component or part of the git repo, you can prefix the title:
+
+```
+CONTRIBUTING.md: Improved instructions for commit messages
+```
+
+This is helpful because it clearly communicates that the change only affects this markdown file.
+The same can be said for only affecting one component(binary):
+
+```
+cf-net: Fixed networking issues on windows
+```
+
+
+### Changelog
+
+Most new feature and fixes should have an entry in the ChangeLog.
+We have an automatic way of generating them before the release, by properly formatting *commit messages*.
+
+* To write arbitrary message in the ChangeLog:
+  `Changelog: <message>`
+* To use the commit title line in the ChangeLog:
+  `Changelog: Title`
+* To use the entire commit message in the ChangeLog:
+  `Changelog: Commit`
+
```

Comment:
I also use Changelog: None to explicitly not have a changelog entry, and have that entry not come up inthe automatic changelog generator. Is that wrong/unnecessary?

Reply:
Yes, please clarify this. I was using this practice as well.

Reply:
Added.